### PR TITLE
MD-5654 Fixed tests for storage plugins

### PR DESCRIPTION
**Pull request by denysord88**
@agozhiy please review.

### CODE REVIEW

File: src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java [hunk 1]
```diff
@@ -30,17 +30,31 @@
 
 public class StoragePage extends BasePage {
 
-  @FindBy(xpath = "//*[text()='Enabled Storage Plugins']/following-sibling::div[@class='table-responsive']")
+  // @FindBy(xpath = "//*[text()='Enabled Storage Plugins']/following-sibling::div[@class='table-responsive']")
+  @FindBy(xpath = "/html/body/div[2]/div[3]")
```
Comment (agozhiy) [inline src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java]:
```suggestion
  @FindBy(xpath = ""//*[text()='Enabled Storage Plugins']/following-sibling::table"")
```

File: src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java [hunk 2]
```diff
@@ -30,17 +30,31 @@
...
   private WebElement disabledStoragePlugins;
 
-  @FindBy(id = "storageName")
-  private WebElement newStoragePluginInput;
+  @FindBy(css = "button[data-target='#new-plugin-modal']")
+  private WebElement newStoragePluginDialog;
 
-  @FindBy(xpath = "//button[text()='Create']")
-  private WebElement newStoragePluginSubmit;
+  @FindBy(xpath = "//*[@id=\"configuration\"]")
```
Comment (agozhiy) [inline src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java]:
```suggestion
  @FindBy(id = "configuration")
```

File: src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java [hunk 3]
```diff
@@ -30,17 +30,31 @@
...
+  private WebElement formTitle;
+
+  @FindBy(css = "input[placeholder='Storage Name']")
+  private WebElement newStoragePluginNameInput;
+
+  @FindBy(xpath = "//*[@id=\"editor\"]/textarea")
+  private WebElement newStoragePluginConfigInput;
+
+  @FindBy(xpath = "//*[@id=\"createForm\"]/div[2]/button[1]")
```
Comment (agozhiy) [inline src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java]:
```suggestion
  @FindBy(id = "createForm")
```
You can find the button further relatively to this element. Don't use indices in locators.

File: src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java [hunk 4]
```diff
@@ -90,6 +104,8 @@ public StoragePage enableStoragePlugin(String name) {
   public StoragePage disableStoragePlugin(String name) {
     WebElement button = getButton(getEnabledStoragePlugins().get(name), "Disable");
     button.click();
+    new WebDriverWait(getDriver(), 5).until(ExpectedConditions.alertIsPresent());
+    AlertSteps.acceptAlert();
```
Comment (agozhiy) [inline src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java]: This shouldn't be in a page object. In general, page doesn't know about steps or tests, such links will create more coupling and do the project more complex. Please avoid this.

File: src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java [hunk 5]
```diff
@@ -99,13 +115,56 @@ public StoragePage exportPlugin(String name) {
...
   }
 
+  public StoragePage openNewStoragePluginDialog() {
+    newStoragePluginDialog.click();
+    return this;
+  }
+
+  public boolean formTitlePresented() {
+    new WebDriverWait(getDriver(), 5)
+        .until(ExpectedConditions.visibilityOf(formTitle));
```
Comment (agozhiy) [inline src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java]: Could you use waitForCondition(Function<WebDriver, Boolean> condition) method instead?

File: src/test/java/org/apache/drillui/test/framework/steps/webui/StorageSteps.java [hunk 6]
```diff
@@ -17,44 +17,65 @@
  */
 package org.apache.drillui.test.framework.steps.webui;
 
-import org.apache.drillui.test.framework.pages.BasePage;
 import org.apache.drillui.test.framework.pages.StoragePage;
 
-public final class StorageSteps {
+import static org.apache.drillui.test.framework.pages.BasePage.getPage;
```
Comment (agozhiy) [inline src/test/java/org/apache/drillui/test/framework/steps/webui/StorageSteps.java:21]: It's better to just create a method getPage(), there is no need to pass the page class every time.

File: src/test/java/org/apache/drillui/test/framework/testng/unsecure/storage/StoragePluginsCRUDTest.java [hunk 7]
```diff
@@ -50,10 +50,6 @@
   @BeforeClass
   public void setupStoragePlugins() {
     RestBaseSteps.setupStoragePlugins();
-  }
-
-  @BeforeMethod
-  public void beforeMethod() {
```
Comment (agozhiy) [inline src/test/java/org/apache/drillui/test/framework/testng/unsecure/storage/StoragePluginsCRUDTest.java]: It's better to do cleanup before every test method.

**Review by agozhiy — CHANGES_REQUESTED** [review]

**Comment by agozhiy** [comment]
Also please clean the obsolete methods from EditStoragePluginSteps:
addPluginMode()
editPluginMode()
create()

**Comment by agozhiy** [comment]
If Firefox there are intermittent failures. They are caused by the insufficient wait conditions: waitForEnabled(), waitForDisabled(). For example, it waits for disappearance of enableButton but doesn't wait for disableButton to appear.

File: src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java (hunk 1 above)
Comment (denysord88) [reply src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java]: Done

File: src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java (hunk 2 above)
Comment (denysord88) [reply src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java]: Done

File: src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java (hunk 3 above)
Comment (denysord88) [reply src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java]: Done

File: src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java (hunk 4 above)
Comment (denysord88) [reply src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java]: Done

File: src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java (hunk 5 above)
Comment (denysord88) [reply src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java]: Done

File: src/test/java/org/apache/drillui/test/framework/steps/webui/StorageSteps.java (hunk 6 above)
Comment (denysord88) [reply src/test/java/org/apache/drillui/test/framework/steps/webui/StorageSteps.java:21]: Done

**Comment by denysord88** [comment]
> If Firefox there are intermittent failures. They are caused by the insufficient wait conditions: waitForEnabled(), waitForDisabled(). For example, it waits for disappearance of enableButton but doesn't wait for disableButton to appear.

Done

File: src/test/java/org/apache/drillui/test/framework/testng/unsecure/storage/StoragePluginsCRUDTest.java (hunk 7 above)
Comment (denysord88) [reply src/test/java/org/apache/drillui/test/framework/testng/unsecure/storage/StoragePluginsCRUDTest.java]: Done

**Comment by denysord88** [comment]
> Also please clean the obsolete methods from EditStoragePluginSteps:
> addPluginMode()
> editPluginMode()
> create()

Done

File: src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java [hunk 8]
```diff
@@ -83,29 +93,80 @@ public EditStoragePluginPage updateStoragePlugin(String name) {
...
-    new WebDriverWait(getDriver(), 5).until(ExpectedConditions.stalenessOf(button));
+    new WebDriverWait(getDriver(), TestProperties.getInt("DEFAULT_TIMEOUT")).until(ExpectedConditions.stalenessOf(button));
     return this;
   }
 
   public StoragePage disableStoragePlugin(String name) {
     WebElement button = getButton(getEnabledStoragePlugins().get(name), "Disable");
     button.click();
-    new WebDriverWait(getDriver(), 5).until(ExpectedConditions.stalenessOf(button));
+    new WebDriverWait(getDriver(), TestProperties.getInt("DEFAULT_TIMEOUT")).until(ExpectedConditions.alertIsPresent());
```
Comment (agozhiy) [inline src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java]:
```suggestion
    waitForCondition(ExpectedConditions.alertIsPresent());
```

File: src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java [hunk 9]
```diff
@@ -83,29 +93,80 @@ public EditStoragePluginPage updateStoragePlugin(String name) {
...
+  public StoragePage setNewStoragePluginConfig(String storageConfig) {
+    sendText(newStoragePluginConfigInput, storageConfig);
+    return this;
+  }
+
+  public StoragePage closeNewPluginForm() {
+    WebElement closeButton = createFormButtons.findElement(By.cssSelector("button[type=button]"));
+    closeButton.click();
+    new WebDriverWait(getDriver(), TestProperties.getInt("DEFAULT_TIMEOUT"))
+        .until(ExpectedConditions.invisibilityOf(closeButton));
```
Comment (agozhiy) [inline src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java]:
```suggestion
        waitForCondition(ExpectedConditions.invisibilityOf(closeButton));
```

File: src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java [hunk 10]
```diff
@@ -83,29 +93,80 @@ public EditStoragePluginPage updateStoragePlugin(String name) {
...
+    return this;
+  }
+
+  public StoragePage submitNewPluginForm() {
+    WebElement submitButton = createFormButtons.findElement(By.xpath("//button[text()='Create']"));
+    submitButton.click();
+    new WebDriverWait(getDriver(), TestProperties.getInt("DEFAULT_TIMEOUT"))
+        .until(ExpectedConditions.invisibilityOf(submitButton));
```
Comment (agozhiy) [inline src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java]:
```suggestion
        waitForCondition(ExpectedConditions.invisibilityOf(submitButton));
```

File: src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java [hunk 11]
```diff
@@ -83,29 +93,80 @@ public EditStoragePluginPage updateStoragePlugin(String name) {
...
+    return this;
   }
-}
+}
```
Comment (agozhiy) [inline src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java]: Add a line feed to the end of the file.

File: src/test/java/org/apache/drillui/test/framework/steps/webui/AuthSteps.java [hunk 12]
```diff
@@ -33,8 +35,21 @@ public static NavigationPage login(String login, String password) {
...
   }
 
+  public static NavigationPage loginFromCustomUrl(String url, String login, String password) {
+    WebBrowser.openURL(url);
+    return BasePage.getPage(ChooseAuthMethodPage.class)
+        .openLoginPage()
+        .setUserName(login)
+        .setUserPassword(password)
+        .submit();
+  }
```
Comment (agozhiy) [inline src/test/java/org/apache/drillui/test/framework/steps/webui/AuthSteps.java]: Is it possible to use one of the login methods into another?

**Review by agozhiy — CHANGES_REQUESTED** [review]

**Comment by agozhiy** [comment]
@denysord88 thank you for the update, just a few minor comments.

File: src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java (hunk 8 above)
Comment (denysord88) [reply src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java]: Done

File: src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java (hunk 9 above)
Comment (denysord88) [reply src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java]: Done

File: src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java (hunk 10 above)
Comment (denysord88) [reply src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java]: Done

File: src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java (hunk 11 above)
Comment (denysord88) [reply src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java]: Done

File: src/test/java/org/apache/drillui/test/framework/steps/webui/AuthSteps.java (hunk 12 above)
Comment (denysord88) [reply src/test/java/org/apache/drillui/test/framework/steps/webui/AuthSteps.java]: Done

**Comment by denysord88** [comment]
Added updates for a confirmation dialog on storage pages.
@agozhiy @Atri0 please review the PR

**Comment by Atri0** [comment]
Looks fine to me

File: src/test/java/org/apache/drillui/test/framework/pages/BasePage.java [hunk 13]
```diff
@@ -152,6 +151,13 @@ protected void forceClick(WebElement element) {
     return result;
   }
 
+  protected void clickOnElement(WebElement element) {
```
Comment (agozhiy) [inline src/test/java/org/apache/drillui/test/framework/pages/BasePage.java]: These two methods: forceClick() and clickOnElement() should be renamed according to their purposes.

File: src/test/java/org/apache/drillui/test/framework/pages/ConfirmDialogPage.java [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+ */
+package org.apache.drillui.test.framework.pages;
+
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+
+public class ConfirmDialogPage extends BasePage {
+  @FindBy(id = "confirmationModal")
+  private WebElement parrentDialogForm;
+  @FindBy(id = "modalBody")
```
Comment (agozhiy) [inline src/test/java/org/apache/drillui/test/framework/pages/ConfirmDialogPage.java:27]: Please separate element definitions by newlines.

File: src/test/java/org/apache/drillui/test/framework/pages/ConfirmDialogPage.java [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+  private WebElement dialogMessage;
+  @FindBy(xpath = "//*[@id='confirmationModal']//*/button")
```
Comment (agozhiy) [inline src/test/java/org/apache/drillui/test/framework/pages/ConfirmDialogPage.java]: There are 3 elements that meet the xpath.

File: src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java [hunk 16]
```diff
@@ -30,17 +31,26 @@
...
 
-  @FindBy(xpath = "//button[text()='Create']")
-  private WebElement newStoragePluginSubmit;
+  @FindBy(id = "configuration")
+  private WebElement formTitle;
+
+  @FindBy(css = "input[placeholder='Storage Name']")
+  private WebElement newStoragePluginNameInput;
+
+  @FindBy(xpath = "//*[@id=\"editor\"]/textarea")
```
Comment (agozhiy) [inline src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java]:
```suggestion
  @FindBy(xpath = "//*[@id='editor']/textarea")
```

**Review by agozhiy — CHANGES_REQUESTED** [review]

File: src/test/java/org/apache/drillui/test/framework/pages/BasePage.java (hunk 13 above)
Comment (denysord88) [reply src/test/java/org/apache/drillui/test/framework/pages/BasePage.java]: Done

File: src/test/java/org/apache/drillui/test/framework/pages/ConfirmDialogPage.java (hunk 14 above)
Comment (denysord88) [reply src/test/java/org/apache/drillui/test/framework/pages/ConfirmDialogPage.java:27]: Done

File: src/test/java/org/apache/drillui/test/framework/pages/ConfirmDialogPage.java (hunk 15 above)
Comment (denysord88) [reply src/test/java/org/apache/drillui/test/framework/pages/ConfirmDialogPage.java]: Done

File: src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java (hunk 16 above)
Comment (denysord88) [reply src/test/java/org/apache/drillui/test/framework/pages/StoragePage.java]: Done

**Review by agozhiy — APPROVED** [review]

**Comment by agozhiy** [comment]
+1, thanks for making the changes.
Please squash the commits and run tests to make sure that everything is ok.

**Pull request merged**